src/anm/analytics.js

```diff
@@ -0,0 +1,79 @@
+var engine = require('engine'),
+    utils = require('./utils.js');
+
+var Analytics = function () {
+    var self = this,
+        supportSendBeacon = !!navigator.sendBeacon,
+        timeout = 1000,
+        beacon = null,
+        animatronUrl = utils.makeApiUrl('/analytics/player');
+
+    self.enabled = animatronUrl != null && animatronUrl.indexOf('animatron-test') >= 0;
+    self.queue = [];
+
+    var event = function () {
+        if (self.queue.length > 0) {
+            var data = JSON.stringify(self.queue);
+            self.queue = [];
+
+            if (supportSendBeacon) {
+                navigator.sendBeacon(animatronUrl, data);
+            } else {
+                sendViaGif(data);
+            }
+        }
+        if (!supportSendBeacon) {
+            setTimeout(event, timeout);
+        }
+    };
+
+    var sendViaGif = function (data) {
+        if (!beacon) {
+            beacon = engine.createStatImg();
+        }
+        beacon.src = animatronUrl + '?data=' + encodeURIComponent(data);
+    };
+    if (self.enabled) {
+        if (!supportSendBeacon) {
+            engine.onDocReady(event);
```

Comment:
Every time I expect my comment will be the last.

```javascript
var events = {}; // could be an array, if you want

function trackEvent(name, stats) {
    if (!events[name]) events[name] = [];
    events[name].push(stats);
}

function getData() {
    return { UID: 'aaa0-bbbx-...', // there's utils.guid()
                 navigator: ...,  events: events };
}

window.addEventListener('unload', function(event) {
  var xhr,
      data = getData();

  if (navigator.sendBeacon) {
    navigator.sendBeacon('api/stats/report', data);
  } else {
    xhr = new XMLHttpRequest(); // there's engine.ajax
    xhr.open('post', '/api/stats/report', false);
    xhr.send(data);
  }
});
```

Here's the code I'd expect here. No timeouts, no intervals. No over-complication. [Source 1](https://w3c.github.io/beacon/), [Source 2](http://www.sitepoint.com/introduction-beacon-api/).